src/messages/proposals.rs

```diff
@@ -202,13 +202,16 @@ impl<'a> QueuedProposal<'a> {
 /// epoch.
 #[derive(Default)]
 pub struct ProposalQueue<'a> {
+    // We keep the references in a separate `Vec` to enforce the correct order
+    proposal_references: Vec<ProposalReference>,
     queued_proposals: HashMap<ProposalReference, QueuedProposal<'a>>,
```

Comment:
The hash map here also holds `ProposalReference`. So I'm not sure I understand the difference. Maybe the struct just needs better comments. Since this is public each member should get a doc comment anyway.